File: finanze/domain/real_estate.py
from datetime import date, datetime
from enum import Enum
from typing import List, Optional
from uuid import UUID

from domain.dezimal import Dezimal
from domain.earnings_expenses import PeriodicFlow
from domain.file_upload import FileUpload
from domain.global_position import InterestType, LoanType
from pydantic.dataclasses import dataclass


class RealEstateFlowSubtype(str, Enum):
    LOAN = "LOAN"
    SUPPLY = "SUPPLY"
    COST = "COST"
    RENT = "RENT"


@dataclass
class LoanPayload:
    type: LoanType
    loan_amount: Optional[Dezimal]
    interest_rate: Dezimal
    euribor_rate: Optional[Dezimal]
    interest_type: InterestType
    fixed_years: Optional[int]
    principal_outstanding: Dezimal
    monthly_interests: Optional[Dezimal] = None


@dataclass
class RentPayload:
    pass


@dataclass
class SupplyPayload:
    tax_deductible: bool = False


@dataclass
class CostPayload:
    tax_deductible: bool = False


RealEstateFlowPayload = LoanPayload | RentPayload | SupplyPayload | CostPayload


@dataclass
class RealEstateFlow:
    periodic_flow_id: Optional[UUID]
    periodic_flow: Optional[PeriodicFlow]
    flow_subtype: RealEstateFlowSubtype
    description: str
    payload: RealEstateFlowPayload


@dataclass
class PurchaseExpense:
    concept: str
    amount: Dezimal
    description: Optional[str]


@dataclass
class Valuation:
    date: date
    amount: Dezimal
    notes: Optional[str]


@dataclass
class Location:
    address: Optional[str] = None
    cadastral_reference: Optional[str] = None


@dataclass
class BasicInfo:
    name: str
    is_residence: bool
    is_rented: bool
    bathrooms: Optional[int] = None
    bedrooms: Optional[int] = None
    photo_url: Optional[str] = None


@dataclass
class PurchaseInfo:
    date: date
    price: Dezimal
    expenses: List[PurchaseExpense]


@dataclass
class ValuationInfo:
    estimated_market_value: Dezimal
    valuations: List[Valuation]
    annual_appreciation: Optional[Dezimal] = None


@dataclass
class Amortization:
    concept: str
    base_amount: Dezimal
    amount: Dezimal
    percentage: Dezimal


@dataclass
class RentalData:
    amortizations: List[Amortization]
    marginal_tax_rate: Optional[Dezimal] = None
    vacancy_rate: Optional[Dezimal] = None


@dataclass
class RealEstate:
    id: Optional[UUID]
    basic_info: BasicInfo
    location: Location
    purchase_info: PurchaseInfo
    valuation_info: ValuationInfo
    flows: List[RealEstateFlow]
    currency: str
    rental_data: Optional[RentalData]
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


@dataclass
class CreateRealEstateRequest:
    real_estate: RealEstate
    photo: Optional[FileUpload]


@dataclass
class UpdateRealEstateRequest:
    real_estate: RealEstate
    remove_unassigned_flows: bool
    photo: Optional[FileUpload]


@dataclass
class DeleteRealEstateRequest:
    id: Optional[UUID]
    remove_related_flows: bool

File: mobile/src/domain/core/realEstate.ts
import { PeriodicFlow } from "./earningsExpenses"
import { FileUpload } from "./fileUpload"
import { InterestType, LoanType } from "./globalPosition"
import { Dezimal } from "@/domain/dezimal"

export enum RealEstateFlowSubtype {
  LOAN = "LOAN",
  SUPPLY = "SUPPLY",
  COST = "COST",
  RENT = "RENT",
}

export interface LoanPayload {
  type: LoanType
  loanAmount: Dezimal | null
  interestRate: Dezimal
  euriborRate: Dezimal | null
  interestType: InterestType
  fixedYears: number | null
  principalOutstanding: Dezimal
  monthlyInterests?: Dezimal | null
}

export interface RentPayload {}

export interface SupplyPayload {
  taxDeductible?: boolean
}

export interface CostPayload {
  taxDeductible?: boolean
}

export type RealEstateFlowPayload =
  | LoanPayload
  | RentPayload
  | SupplyPayload
  | CostPayload

export interface RealEstateFlow {
  periodicFlowId: string | null
  periodicFlow: PeriodicFlow | null
  flowSubtype: RealEstateFlowSubtype
  description: string
  payload: RealEstateFlowPayload
}

export interface PurchaseExpense {
  concept: string
  amount: Dezimal
  description: string | null
}

export interface Valuation {
  date: string
  amount: Dezimal
  notes: string | null
}

export interface Location {
  address?: string | null
  cadastralReference?: string | null
}

export interface BasicInfo {
  name: string
  isResidence: boolean
  isRented: boolean
  bathrooms?: number | null
  bedrooms?: number | null
  photoUrl?: string | null
}

export interface PurchaseInfo {
  date: string
  price: Dezimal
  expenses: PurchaseExpense[]
}

export interface ValuationInfo {
  estimatedMarketValue: Dezimal
  valuations: Valuation[]
  annualAppreciation?: Dezimal | null
}

export interface Amortization {
  concept: string
  baseAmount: Dezimal
  amount: Dezimal
  percentage: Dezimal
}

export interface RentalData {
  amortizations: Amortization[]
  marginalTaxRate?: Dezimal | null
  vacancyRate?: Dezimal | null
}

export interface RealEstate {
  id: string | null
  basicInfo: BasicInfo
  location: Location
  purchaseInfo: PurchaseInfo
  valuationInfo: ValuationInfo
  flows: RealEstateFlow[]
  currency: string
  rentalData: RentalData | null
  createdAt?: string | null
  updatedAt?: string | null
}

export interface CreateRealEstateRequest {
  realEstate: RealEstate
  photo: FileUpload | null
}

export interface UpdateRealEstateRequest {
  realEstate: RealEstate
  removeUnassignedFlows: boolean
  photo: FileUpload | null
}

export interface DeleteRealEstateRequest {
  id: string | null
  removeRelatedFlows: boolean
}
